Pattern input field

brunopagno opened the pull request:
# Ticket

https://community.openproject.org/projects/document-workflows-stream/work_packages/59999

Strongly related to https://github.com/opf/openproject/pull/17472

## Screenshots

https://github.com/user-attachments/assets/95a01058-9184-43a0-acfc-74e0125a5b58

# What approach did you choose and why?

This is a bit of an alternative version to the original design. The input field behaves as a "tiny text editor" with autocomplete options and inserts editable tokens onto it's body.

- We do not validate tokens *in this PR*, so they can be submitted with invalid values

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

brunopagno committed:
create a pattern autocompleter input

This commit introduces a new type of input field which can handle autocomplete with 'tokens', so that we can build
patterns like `Example pattern {{token}} value`.

lindenthal commented:
Hi @dominic-braeunlein,
I just tried to create a pullpreview environment. The build seems not to be started. Can you please check?
Thanks
Niels

dominic-braeunlein commented:
> Hi @dominic-braeunlein, I just tried to create a pullpreview environment. The build seems not to be started. Can you please check? Thanks Niels

FYI: Pull-preview is deployed

@lindenthal

brunopagno committed:
make form control look like a form control

brunopagno committed:
check for invalid tags on load

Kharonus commented on frontend/src/stimulus/setup.ts:
```diff
@@ -14,6 +14,7 @@ import OpShowWhenValueSelectedController from './controllers/show-when-value-sel
 import FlashController from './controllers/flash.controller';
 import OpProjectsZenModeController from './controllers/dynamic/projects/zen-mode.controller';
 import PasswordConfirmationDialogController from './controllers/password-confirmation-dialog.controller';
+import PatternInputController from './controllers/pattern-input.controller';
```
🟡 what is the reasoning behind having a fixed import vs. using the dynamic controller?

brunopagno replied:
I could not get the automatic one to work. But I can check again.

brunopagno replied:
I think if we have several "dynamic" controllers on the same page it does not work properly. That or the fact that this component is a Form one, not an ApplicationComponent

Kharonus commented on app/forms/work_packages/types/subject_configuration_form.rb:
```diff
@@ -51,12 +51,13 @@ class SubjectConfigurationForm < ApplicationForm
-          toggleable_group.text_field(
+          toggleable_group.pattern_input(
             name: :pattern,
+            value: model.pattern,
```
❓ is that needed? usually the form mapping should work, if the name is equal to the attribute of the model.

Kharonus commented on app/components/work_packages/types/pattern_input.html.erb:
```diff
@@ -0,0 +1,93 @@
+  </template>
+
+  <template data-pattern-input-target="suggestionsHeadingTemplate">
+    <%= render(Primer::Alpha::ActionList::Heading.new(title: "__placeholder__", heading_level: 2)) %>
```
🟡 What is this `__placeholder__` string happening here in multiple places?

brunopagno replied:
You can notice that these are `templates`  targeted by the stimulus controller. The idea is that these have placeholder values that will be replaced by JS dynamically.

I bet there's a better pattern, but this felt simple and explicit enough. If you feel like there's another path I'm listening

Kharonus commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -0,0 +1,288 @@
+    this.suggestionsTarget.innerHTML = '';
+  }
+
+  private filterSuggestions(word:string):void {
```
🟡 complexity of this method is too high for me.

Can we extract the construction of the `filtered` array into a separate method?

Kharonus commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -0,0 +1,288 @@
+    // insert the HTML
+    filtered.forEach((group) => {
+      const groupHeader = this.suggestionsHeadingTemplateTarget.content?.cloneNode(true) as HTMLElement;
+      groupHeader.querySelector('h2')!.innerText = group.key;
```
🔴 Null assertion in type script is usually considered a bad style.

Can you extract the assignment to safer method, which raises an explanatory error if something is off and expectations are not met.

brunopagno replied:
I meant to do this. Thanks for catching it :bow:

HDinger replied:
This block still has the null assertion in it which is doubled dangerous as the tag you are searching for could change and the `groupHeader` itself could be nil. Please catch these cases

Kharonus commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -0,0 +1,288 @@
+      const token = element.textContent?.trim();
+
+      let exists = false;
+      Object.keys(this.suggestionsInitialValue).forEach((key) => {
```
🟡 As you are searching for the values here, you might want to directly iterate over `Object.values()`, which will give you directly the `group`.

Kharonus commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -0,0 +1,288 @@
+    });
+  }
+
+  private tagInvalidTokens():void {
```
🟡 having this executed over and over makes me wonder, if we maybe want to build another flat list of all tokens on `connect`, so that the contained check is simply `allTokens.includes(token)`.

brunopagno replied:
To be really honest I don't think this one is too heavy lifting :sweat_smile:.

But you do have a good point. Lemme try!

Kharonus commented on frontend/src/stimulus/controllers/pattern-input.controller.ts:
```diff
@@ -0,0 +1,288 @@
+    this.clearSuggestionsFilter();
+
+    const filtered = Object.keys(this.suggestionsInitialValue).map((key) => {
+      const group = this.suggestionsInitialValue[key];
```
🟡 if you need both, the `key` and the `value` use `Object.entries` as an iterator.

`Object.entries(this.suggestionsInitialValue).map(([key, group]) => { /*...*/}`

brunopagno committed:
code requested changes

HDinger reviewed (changes requested):
Hi @brunopagno 

Nice work 👍  This looks so much better than the first iteration and there are much less issues. I checked Safari, Firefox and Chrome and there was no difference to be seen between the browsers.  👏 🥳 

I did found some issues while using it, whereas I think the first one is the most important one:

- [x] When selecting an item, I cannot leave the token any more. Sometimes that also happens at the beginning of the field: I chose a token and cannot add anything to the left. 

![Jan-29-2025 16-41-35](https://github.com/user-attachments/assets/bb732ff2-625f-49c0-8e33-e9689120e48e)

- [x] The field is too long compared to the visuals in Figma

- [x] There are some styling issues on the right of the list: The separator overflows, while the selction has no space to the right

<img width="851" alt="Bildschirmfoto 2025-01-29 um 17 00 48" src="https://github.com/user-attachments/assets/7f53b927-4221-4f97-a221-585e42ca3edc" />

- [ ] I can save an invalid state. I assume that this is just treated as text, but it looks like an invalid token:
<img width="244" alt="Bildschirmfoto 2025-01-29 um 17 03 54" src="https://github.com/user-attachments/assets/583f7f46-560b-41b2-91ca-3bb3ac9f3245" />

- [x] When I delete a token, the input field does not switch back to "normal" mode. The font is still red, and as soon as I select anything, the whole content is replaced by that selection:
![Jan-29-2025 17-06-07](https://github.com/user-attachments/assets/182f3425-9038-4884-bfc8-6c34276ac9b9)

- [ ] When I am in an existing token, I sometimes manage to get results which do not change anything:
![Jan-29-2025 17-11-12](https://github.com/user-attachments/assets/20eb682f-a75f-45a6-9410-8c2c8e58e845)

- [x] I don't know what we said about whether the user should be drag items around? Currently this is not possible, but maybe that is expected.

- [x] On page load, I can see the pattern_input component although "manual" is selected before it vanishes again. This is however probably out of scope for this PR I assume?
<img width="565" alt="Bildschirmfoto 2025-01-29 um 17 14 28" src="https://github.com/user-attachments/assets/e28cbafc-6850-4d05-88db-0d129366d4cd" />

- [ ] last but not least: the component is not accessible for screen readers. It is not read that an input appears when selecting the option, it is not said that an option list appeared, and the token transoformation and state is also not read. Currently this is an admin only component, which might need less focus on Accessiblity. However, if we want to use that on other places, it probably has to be accessible. We can discuss this if you want. Maybe @akabiru also has some ideas.

brunopagno committed:
handle case when the cursor gets stuck inside a token

brunopagno commented:
> When selecting an item, I cannot leave the token any more. Sometimes that also happens at the beginning of the field: I chose a token and cannot add anything to the left.

I implemented a solution for this.

> The field is too long compared to the visuals in Figma
> There are some styling issues on the right of the list: The separator overflows, while the selction has no space to the right

I will push these for later improvements.

> I can save an invalid state. I assume that this is just treated as text, but it looks like an invalid token

This is known and we will handle validation separately. Not in scope for this PR.

> When I delete a token, the input field does not switch back to "normal" mode. The font is still red, and as soon as I select anything, the whole content is replaced by that selection

Chromium has some weirdness that it tries to keep the style on the cursor. So it includes a <font> tag to the DOM which effectively break some functionality of the component.

https://stackoverflow.com/questions/14145970/chrome-empty-contenteditable-retaining-styling

I believe I fixed it.

> When I am in an existing token, I sometimes manage to get results which do not change anything

I could not reproduce. It might have been related to the item above inserting alien tags.

> I don't know what we said about whether the user should be drag items around? Currently this is not possible, but maybe that is expected.

As far as I understood we do not intend to handle dragging items. So I will say this is expected.

> On page load, I can see the pattern_input component although "manual" is selected before it vanishes again. This is however probably out of scope for this PR I assume?

Not in scope. This is more likely related to loading the form + connecting stimulus. Not really responsibility of the component.

> last but not least: the component is not accessible for screen readers...

Understood. I will follow up with the team.

brunopagno committed:
handle chrome retaining style

brunopagno commented:
@HDinger @Kharonus any chance one of you can take another look at this one today? I appreciate a lot the help :bow:

HDinger reviewed (commented):
Works like charm, good job 👍 🌟 There are some issues with regards to styling:

* The input is still too long
* The separators are overflowing the dropdown.

These are no blockers for the merge, but need to be tackled before the release. I am fine howover with tackling that separately. Please also remeber to document the component in lookbook and follow-up on the accessibility.

brunopagno committed:
handle null assertion elegantly

brunopagno commented:
@HDinger Thanks. And agreed. It makes sense, I'll follow up on a separate PR.

Also I fixed the code change requested.

HDinger reviewed (approved):
Good job @brunopagno 👏
